packages/debug/src/browser/debug-session-manager.ts

```diff
@@ -287,6 +287,16 @@ export class DebugSessionManager {
         return this.start(options);
     }
 
+    async terminateSessions(): Promise<void> {
+        this.updateCurrentSession(undefined);
+        this.currentSession?.terminate();
+    }
+
+    async restartSessions(): Promise<void> {
+        this.updateCurrentSession(undefined);
+        this.currentSession?.restart();
+    }
```

Comment:
The naming of the methods are misleading, we aren't dealing with multiple sessions.

Reply:
By those before me that defined debug sessions logic - each item in `Threads` section is considered a Session in Theia - therefore I named it as such. How would you like me to name them?

Reply:
@danarad05 in the case we have multiple ongoing debug sessions, does this logic terminate them all (even ones which are unrelated)?

Reply:
@vince-fugnitto 
1. didn't think about that possibility. Is that a common usage - running multiple debug "launches"? if that is needed we might want to open issue for separating those n session groups from each others - in UI - as currently you cannot distinguish between them - they are all listed as equal.
2. I tested it out a little - terminating on "stop" icon (without selection a session) - terminates first session and it's child sessions. So if there are 2 different session "launches" - it terminates sessions of first group leaving second group intact. Still, there are caveats  - for instance you cannot run 2 pwa-extensionHost launches as currently extension is run on localhost:3030 (hardcoded 3030). so I wouldn't say it is full proof but probably good enough for this PR currently.

Reply:
@danarad05 one example I know of in our own repo is starting two configurations as the following:
- start browser backend
- start browser frontend

The debug-view has two sessions in this instance (note that vscode does not have the `threads` view):

![image](https://user-images.githubusercontent.com/40359487/111339659-52d20a00-864e-11eb-870a-49d8f2f1f517.png)

Reply:
ok. 
First: that I guess is another issue - that Theia doesn't support "compound" launch configs like vscode.

Second: tried running both launch BE and FE which opened localhost:3000 in a new browser and worked (for the most part - as it was pretty slow and stuck a little using it). I also stop/restarted and again - it did it but I not perfectly (new `session` was added to end of list.

first run of both sessions:
![image](https://user-images.githubusercontent.com/71069170/111441125-b112f180-870f-11eb-8db6-f14167c68b85.png)

After restarting first session - a new session was added last:
![image](https://user-images.githubusercontent.com/71069170/111441311-e15a9000-870f-11eb-93a2-0dc8dad60899.png)

Third: breakpoints unfortunately in parent localhost:3000 weren't hit when running child localhost:3000.

So, I wouldn't say it's all perfect but maybe it's time to merge this PR and move forward with it.

Reply:
> So, I wouldn't say it's all perfect but maybe it's time to merge this PR and move forward with it.

@danarad05 I'm just trying to understand potential implications to the debug side of the changeset and how `termination` of a debug session is now handled. Looking at VS Code's sources, they handle stopping a session like so:


```ts
async stopSession(session: IDebugSession | undefined): Promise<any> {
	if (session) {
		return session.terminate();
	}

	const sessions = this.model.getSessions();
	if (sessions.length === 0) {
		this.taskRunner.cancel();
		// User might have cancelled starting of a debug session, and in some cases the quick pick is left open
		await this.quickInputService.cancel();
		this.endInitializingState();
		this.cancelTokens(undefined);
	}

	return Promise.all(sessions.map(s => s.terminate()));
}
```

It seems in this instance, the `stop` toolbar command terminates all sessions?

- [stopHandler](https://github.com/microsoft/vscode/blob/b335d9eaf100990547ae45f3fdac54ae129dcbc3/src/vs/workbench/contrib/debug/browser/debugCommands.ts#L281-L298)
- [stopSession](https://github.com/microsoft/vscode/blob/b335d9eaf100990547ae45f3fdac54ae129dcbc3/src/vs/workbench/contrib/debug/browser/debugService.ts#L739-L754)

Reply:
@vince-fugnitto 
Seems like:
1. In VS Code there is no debug toolbar with 'play', 'stop', 'restart' commands like in Theia and commands are on "sessions" themselves:
![image](https://user-images.githubusercontent.com/71069170/111479383-981f3600-8739-11eb-96c8-5d65b43924fa.png)

2. I see also that `stop` command exist on parent sessions only (first one created upon running debug).  
![image](https://user-images.githubusercontent.com/71069170/111479832-0663f880-873a-11eb-9a97-04015adc96ed.png)

3. looking at the code you noted, stopping invokes stop on parent session - which is similar to what I did (going for first session (= sessions[0]) when user doesn't select session from Threads section) while in Theia we didn't implement "parentSession" connection between sessions.
```
// Stop should be sent to the root parent session
	while (!showSubSessions && session && session.parentSession) {
		session = session.parentSession;
	}

	await debugService.stopSession(session);
```

Reply:
> 1. In VS Code there is no debug toolbar with 'play', 'stop', 'restart' commands like in Theia and commands are on "sessions" themselves:

That's not entirely true, I do see the toolbar items (I use the `docked` location - `"debug.toolBarLocation": "docked"`):

![image](https://user-images.githubusercontent.com/40359487/111481615-845bdb80-8709-11eb-96e3-27d2def9ac47.png)

> 3. looking at the code you noted, stopping invokes stop on parent session - which is similar to what I did (going for first session (= sessions[0]) when user doesn't select session from Threads section) while in Theia we didn't implement "parentSession" connection between sessions.

For my own understanding, is `sessions[0]` guaranteed to be the parent session?

Reply:
> That's not entirely true, I do see the toolbar items (I use the docked location - "debug.toolBarLocation": "docked"):

You are right. my bad, it was hidden... 

> For my own understanding, is sessions[0] guaranteed to be the parent session?

Yes, sessions[0] is parent session for first launch as it is first session created.